### PR TITLE
Add `X-NUCLEO-IDW01M1` expansion board

### PR DESCRIPTION
**Pull request by betzw**
cc @MarceloSalazar 

See https://github.com/ARMmbed/wifi-x-nucleo-idw01m1/issues/1#issuecomment-323332087

### CODE REVIEW

**Comment by MarceloSalazar** [comment]
@geky @bulislaw you may want to review this, as other wifi modules may need to be added here as well.

File: README.md [hunk 1]
```diff
@@ -12,50 +12,62 @@ The program brings up the WiFi and the underlying network interface, and uses it
 
 * [UBLOX Odin board](https://developer.mbed.org/platforms/ublox-EVK-ODIN-W2/) built-in WiFi module
 * [REALTEK_RTL8195AM](https://developer.mbed.org/platforms/REALTEK-RTL8195AM/) built-in WiFi module
+* [NUCLEO_F401RE](https://developer.mbed.org/platforms/ST-Nucleo-F401RE/) with [X-NUCLEO-IDW01M1](https://developer.mbed.org/components/X-NUCLEO-IDW01M1/) WiFi expansion board using pins D8 D2
 * [NUCLEO_F429ZI](https://developer.mbed.org/platforms/ST-Nucleo-F429ZI/) with ESP8266-01 module using pins D1 D0
```
Comment (betzw) [inline README.md]: cc @nikapov-ST

File: README.md [hunk 2]
```diff
@@ -12,50 +12,62 @@ The program brings up the WiFi and the underlying network interface, and uses it
...
+               "value": "\"SSID\""
+           },
+           "wifi-password": {
+               "help": "WiFi Password",
+               "value": "\"Password\""
+           }
+       },
+   ```
+
+3. Copy the respective _ignore_ file to `.mbedignore`
```
Comment (bulislaw) [inline README.md]: I don't like that, why can't we compile both and let the linker strip the symbols from the other one? Do we get some sort of name clashes?

**Comment by bulislaw** [comment]
@kjbracey-arm @JanneKiiskila any thoughts?

File: README.md (hunk 1 above)
Comment (nikapov-ST) [inline README.md]: @MarceloSalazar I have just updated the component page adding an explicit link to this repo. Pls. have a look and let me know.

Comment (MarceloSalazar) [inline README.md]: Looks good. Thanks.

**Comment by betzw** [comment]
I think both drivers use the `ATParser` which might lead to name clashes.

**Comment by kjbracey-arm** [comment]
Drivers should be transitioning to use the updated and renamed in-tree ATCmdParser and UARTSerial,

The original ATParser does also still exist as a separate repo, and we ensured no name clashes betweeen those two.

But there shouldn't be any other copies of ATParser kicking about in drivers' own repos any more.

**Comment by betzw** [comment]
- Is `ATCmdParser` backward compatible with `ATParser`? 
- Both drivers do **not** bring in there own copy of `ATParser` But refer to the separate repo of the original `ATParser` you mentioned above. 
- Who and when should the separate repo of the original `ATParser` be used?

**Comment by betzw** [comment]
I have taken a quick look at the in-tree classes `ATCmdParser` and `UARTSerial` and it seems that `ATCmdParser` is (still) lacking the distinction between send & receive delimiters.

**Comment by kjbracey-arm** [comment]
AtCmdParser autodetects line endings on receive, borrowing code from retarget.cpp for stdin - it should adapt to whatever the device outputs.

Therefore the delimiter setting only affects output.

**Comment by kjbracey-arm** [comment]
In answer to the earlier questions, AtCmdParser is largely but not totally compatible. Main difference is that it can in principle attach to any comms device via the generic FileHandle interface, rather than having its own specific BufferedSerial built-in, so construction and attachment differs, but a few improvements and fixes were made while working on it, such as that RX delimiter auto-handling.

There's no reason to use the external ATParser in new code - that remains there for existing components. And that, as you note, has the potential of clashes if two drivers each pull in that lib.  (I don't believe the build system can detect duplicate .lib inclusions)

**Comment by betzw** [comment]
Is there a way to override the values of `uart-serial-rxbuf-size` & `uart-serial-txbuf-size` at library level, i.e. in file `mbed_lib.json` of the wifi driver, rather than just at application level using `mbed_app.json`?

**Comment by kjbracey-arm** [comment]
I'm not an expert on the config system, but I don't believe there's currently a way for a library to request a config change in another library. Each setting can only be set once, by the application, else it's the default. (Unless someone else knows differently?)

For other purposes, I've previously thought it would be generally useful to allow this - have the final setting of a numerical value be either the sum or maximum of all requests from various components.

I would have preferred to have buffer size be a run-time setting as a constructor parameter, but UARTSerial inherited the template-sized circular buffer from the original BufferedSerial, and I didn't really want to make UARTSerial itself be a template, and changing the buffering wasn't a priority.

**Comment by betzw** [comment]
Well, at that point I do not think it makes currently sense to move WiFi drivers (e.g. ESP8266, IDW01M1) from `ATParser` to `ATCmdParser` and from `BufferedSerial` to `UARTSerial`.
So I would suggest to stick with this PR as it is, at least temporarily until this issue has not been fixed.

cc @MarceloSalazar, @screamerbg, @nikapov-ST

**Comment by kjbracey-arm** [comment]
I agree that there's no need to immediately move necessarily - but if you did have some sort of duplicate code problem in this PR due to the twin ATParser as mentioned above, it would avoid it. 

There are no issues I'm aware of preventing transitioning any driver - any limitations of UARTSerial/AtCmdParser such as app-configured buffered size already exist in BufferedSerial/ATParser - they're just the same thing revamped and renamed to avoid compatibility grief with the external components.

**Comment by betzw** [comment]
> any limitations of UARTSerial/AtCmdParser such as app-configured buffered size already exist in BufferedSerial/ATParser 

As far as I understand in `BufferedSerial`/`ATParser` the buffer sizes are past as parameters in the constructor of `BufferedSerial` and therefore are currently decided by the WiFi driver and not at application level.

**Comment by kjbracey-arm** [comment]
My mistake - I was somehow sure that the use of that CircularBuffer was inherited from BufferedSerial. But it was substituted for "MyBuffer" because it was available in-tree, adding this limitation :(

I would normally expect the default 256 bytes to be enough though - ATParser needs to contain a line-sized buffer, but (or so?) there should be no need to allocate another line-sized buffer in the serial driver. Data should easily stream through the driver reliably with a 256-byte buffer at 115200 baud.

The PPP-based and AT-command-based cellular drivers using UARTSerial have so far not needed to raise the buffer size, except for some experiments at 460800 and above which found 512 beneficial. I can't immediately see why you'd need the 2K in your driver.

**Comment by betzw** [comment]
On my side the TX buffer tends to overflow regularly.

**Comment by screamerbg** [comment]
@kjbracey-arm What's gating this PR? @betzw @nikapov-ST and the rest of the ST team are following the reference implementation in the ESP8266 driver.

If you feel that the ST team should follow a different WiFi reference implementation then please provide link to it and/or update the ESP8266 so there's less confusion what's the "right" way. If not, then let's progress with this PR and get the shield supported in this example program.

**Comment by screamerbg** [comment]
Bump @kjbracey-arm @MarceloSalazar. See my comment above.

File: mbed_app.json [hunk 3]
```diff
@@ -1,5 +1,9 @@
 {
     "config": {
+	"wifi-shield": {
+            "help": "Options are WIFI_ESP8266, WIFI_IDW01M1",
+            "value": "WIFI_IDW01M1"
+	},
         "wifi-ssid": {
```
Comment (MarceloSalazar) [inline mbed_app.json]: @betzw Can we add an entry in the 'target_overrides' section in this file to get the NUCLEO_F401RE working out of the box with D8, D2?

File: README.md (hunk 2 above)
Comment (MarceloSalazar) [reply README.md]: Indeed, there is a clash at link time, and that's why the need for the `.mbedignore`

File: README.md [hunk 4]
```diff
@@ -12,50 +12,62 @@ The program brings up the WiFi and the underlying network interface, and uses it
...
+   * Copy `esp8266-mbedignore` when using ESP2866 module.
```
Comment (MarceloSalazar) [inline README.md]: ESP2866 --> ESP8266

**Comment by MarceloSalazar** [comment]
@screamerbg the ESP8266 is a reference example on how to implement the wifi driver; until that is updated, I suggest to continue using the external ATParser.

**Comment by MarceloSalazar** [comment]
@betzw I'm not able to get the example working with the F401 + X-NUCLEO-IDW01M1 wifi module.
The app gets stack in ("Connecting...")[https://github.com/ARMmbed/mbed-os-example-wifi/blob/master/main.cpp#L134]
Is there anything else needed on your betzw:betzw_wb repo?

File: mbed_app.json (hunk 3 above)
Comment (betzw) [reply mbed_app.json]: Done.

File: README.md (hunk 4 above)
Comment (betzw) [reply README.md]: Done.

**Comment by betzw** [comment]
Other than configuring `mbed_app.json` for your environment etc. there should be nothing else special to be done.
Having the module not being able to connect is typically due to a typo in the SSID or password, but I am sure that you have more than double checked.
I have also seen that the sha-id in `wifi-x-nucleo-idw01m1.lib` was outdated, maybe this is the problem?

**Comment by MarceloSalazar** [comment]
Confirmed it works now

**Pull request merged**